orchagent/pbh/pbhmgr.cpp

```diff
@@ -0,0 +1,1109 @@
+    { PBH_HASH_FIELD_HASH_FIELD_INNER_DST_IPV6,    SAI_NATIVE_HASH_FIELD_INNER_DST_IPV6    },
+    { PBH_HASH_FIELD_HASH_FIELD_INNER_SRC_IPV6,    SAI_NATIVE_HASH_FIELD_INNER_SRC_IPV6    }
+};
+
+// functions ----------------------------------------------------------------------------------------------------------
+
+template<typename T>
+static inline T toUInt(const std::string &hexStr)
+{
+    if (hexStr.substr(0, 2) != "0x")
+    {
+        throw std::invalid_argument("Invalid argument: '" + hexStr + "'");
+    }
+
+    return to_uint<T>(hexStr);
+}
+
+static inline std::uint8_t toUInt8(const std::string &hexStr)
+{
+    return toUInt<std::uint8_t>(hexStr);
+}
+
+static inline std::uint16_t toUInt16(const std::string &hexStr)
+{
+    return toUInt<std::uint16_t>(hexStr);
+}
+
+static inline std::uint32_t toUInt32(const std::string &hexStr)
+{
+    return toUInt<std::uint32_t>(hexStr);
+}
+
+// PBH helper ---------------------------------------------------------------------------------------------------------
+
+bool PbhHelper::hasDependencies(const PbhContainer &obj) const
+{
+    return obj.getRefCount() > 0;
+}
+
+template<>
+bool PbhHelper::validateDependencies(const PbhRule &obj) const
+{
+    const auto &tCit = this->tableMap.find(obj.table);
+    if (tCit == this->tableMap.cend())
+    {
+        return false;
+    }
+
+    const auto &hCit = this->hashMap.find(obj.hash.value);
+    if (hCit == this->hashMap.cend())
+    {
+        return false;
+    }
+
+    return true;
+}
+
+template<>
+bool PbhHelper::validateDependencies(const PbhHash &obj) const
+{
+    for (const auto &cit : obj.hash_field_list.value)
+    {
+        const auto &hfCit = this->hashFieldMap.find(cit);
+        if (hfCit == this->hashFieldMap.cend())
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+template<>
+bool PbhHelper::addDependencies(const PbhRule &obj)
+{
+    const auto &tCit = this->tableMap.find(obj.table);
+    if (tCit == this->tableMap.cend())
+    {
+        return false;
+    }
+
+    const auto &hCit = this->hashMap.find(obj.hash.value);
+    if (hCit == this->hashMap.cend())
+    {
+        return false;
+    }
+
+    auto &table = tCit->second;
+    table.incrementRefCount();
+
+    auto &hash = hCit->second;
+    hash.incrementRefCount();
+
+    return true;
+}
+
+template<>
+bool PbhHelper::addDependencies(const PbhHash &obj)
+{
+    std::vector<std::unordered_map<std::string, PbhHashField>::iterator> itList;
+
+    for (const auto &cit : obj.hash_field_list.value)
+    {
+        const auto &hfCit = this->hashFieldMap.find(cit);
+        if (hfCit == this->hashFieldMap.cend())
+        {
+            return false;
+        }
+
+        itList.push_back(hfCit);
+    }
+
+    for (auto &it : itList)
+    {
+        auto &hashField = it->second;
+        hashField.incrementRefCount();
+    }
+
+    return true;
+}
+
+template<>
+bool PbhHelper::removeDependencies(const PbhRule &obj)
```

Comment:
As far as I see the purpose of this function is to decrementAllDependencyRefCountsIs and not removeDependencies, I would recommend changing it per purpose of the function.

Reply:
Similar comment applies to other types of Pbh objects as well

Reply:
@anish-n what other types of PBH objects? Can you please elaborate?

As for your comment we can do:
```cpp
bool hasDependencies(const PbhContainer &obj) const;
template<typename T>
bool validateDependencies(const T &obj) const;

template<typename T>
bool incRefCount(const T &obj);
template<typename T>
bool decRefCount(const T &obj);
```

Is that what you are expecting?